numademo.c

```diff
@@ -436,7 +436,7 @@ void test(enum test type)
 		numa_set_localalloc();
 		memtest("local allocation", numa_alloc(msize));
 
-		numa_set_preferred((node_to_use[i]+1) % numnodes );
```

Comment:
This code preceded `memtest("setting wrong preferred node", numa_alloc(msize));`, but after your change it behaves just the same as `memtest("setting wrong preferred node", numa_alloc(msize));`...

In a way, this is trying to find the "next" node to create a mismatch.

Perhaps a possible approach would be earlier on when listing valid nodes create a map that points to the next valid node from the current one?

Reply:
Perhaps this might work?

```c
numa_set_preferred(node_to_use[(i + 1) % numnodes]);
```

That assumes that above in `get_node_list()`, `got_nodes` in the end matches `numnodes`, perhaps adding an assertion for that would be good there...

Reply:
@filbranden The actual trick I missed earlier and this fix is the precedence with which `i` was used. As you pointed out, `(node_to_use[(i + 1) % numnodes])` rightly provides "wrongly preferred node". And yes, `got_nodes` should match the `numnodes`, i have added a check for it.